src/executor.rs

```diff
@@ -0,0 +1,142 @@
+use std::mem::replace;
+use pyo3::PyResult;
+use revm::{Context, ContextWithHandlerCfg, Evm, EvmContext, FrameOrResult, FrameResult, inspector_handle_register};
+use revm::inspectors::TracerEip3155;
+use revm::primitives::{ExecutionResult};
+use revm::primitives::TransactTo;
+use revm_interpreter::{CallInputs, CreateInputs, SuccessOrHalt};
+use revm_interpreter::primitives::HandlerCfg;
+use crate::database::DB;
+use crate::utils::pyerr;
+
+pub(crate) fn call_evm(evm_context: EvmContext<DB>, handler_cfg: HandlerCfg, tracing: bool) -> PyResult<(ExecutionResult, EvmContext<DB>)> {
+    let (result, evm_context) = if tracing {
+        let tracer = TracerEip3155::new(Box::new(crate::pystdout::PySysStdout {}), true);
+        let evm = Evm::builder()
+            .with_context_with_handler_cfg(ContextWithHandlerCfg {
+                cfg: handler_cfg,
+                context: Context {
+                    evm: evm_context,
+                    external: tracer,
+                },
+            })
+            .append_handler_register(inspector_handle_register)
+            .build();
+        evm_call(evm)
+    } else {
+        let evm = Evm::builder()
+            .with_context_with_handler_cfg(ContextWithHandlerCfg {
+                cfg: handler_cfg,
+                context: Context {
+                    evm: evm_context,
+                    external: (),
+                },
+            })
+            .build();
+
+        evm_call(evm)
+    }?;
+    Ok((result, evm_context))
+}
+
+fn evm_call<EXT>(mut evm: Evm<'_, EXT, DB>) -> PyResult<(ExecutionResult, EvmContext<DB>)> {
+    evm.handler.validation().env(&evm.context.evm.env).map_err(pyerr)?;
+    let initial_gas_spend = evm
+        .handler
+        .validation()
+        .initial_tx_gas(&evm.context.evm.env)
+        .map_err(pyerr)?;
+    evm.handler
+        .validation()
+        .tx_against_state(&mut evm.context)
+        .map_err(pyerr)?;
+
+    let ctx = &mut evm.context;
+    let pre_exec = evm.handler.pre_execution();
+
+    // load access list and beneficiary if needed.
+    pre_exec.load_accounts(ctx).map_err(pyerr)?;
+
+    // load precompiles
+    ctx.evm.set_precompiles(pre_exec.load_precompiles());
+
+    // deduce caller balance with its limit.
+    pre_exec.deduct_caller(ctx).map_err(pyerr)?;
+
+    let gas_limit = ctx.evm.env.tx.gas_limit - initial_gas_spend;
+
+    let exec = evm.handler.execution();
+    // call inner handling of call/create
+    let first_frame_or_result = match ctx.evm.env.tx.transact_to {
+        TransactTo::Call(_) => exec.call(
```

Comment:
i think it will probably be cleaner do have two different functions `call` and `create` instead of using the `tx.transact_to` to parametrize behavior

Reply:
but, maybe it's cleaner this way. i'm not sure :thinking:

Reply:
I'm assuming this is mostly copied from revm, is there a reason for this? Also cc @rakita

Reply:
It's [here](https://github.com/gakonst/pyrevm/blob/618ef67d6a2a4541288ea3ac91de17a19b040736/src/executor.rs#L41) and [here](https://github.com/gakonst/pyrevm/blob/618ef67d6a2a4541288ea3ac91de17a19b040736/src/executor.rs#L105)

Reply:
https://github.com/gakonst/pyrevm/blob/618ef67d6a2a4541288ea3ac91de17a19b040736/src/executor.rs#L41
https://github.com/gakonst/pyrevm/blob/618ef67d6a2a4541288ea3ac91de17a19b040736/src/executor.rs#L105

Reply:
In Evm sense, this is how call/create are specified and are differentiated by that field, so it is best to stay like that. There is one TX type that can be CALL or CREATE.